tensorflow/core/util/mkl_util.h

```diff
@@ -1755,7 +1757,76 @@ class MklDnnData {
   }
 };
 
-#endif  // INTEL_MKL_ML
+/// Base class for operations with reuse of DNN primitives
+///
+class DnnOp {
+ public:
+  virtual ~DnnOp() {}
+  virtual void Setup() {return;}
+  virtual void Execute() {return;}
+
+  // dummy data
+  unsigned char DummyData[256];
+};
+
+const mkldnn::memory::dims NONE_DIMS = {};
+// This constant is used to declare dummy buffer (size), for MKL primitives
+template <typename T>
+class DnnOpFactory {
+ public:
+  DnnOpFactory() {}
+  ~DnnOpFactory() {}
+
+  DnnOp* GetOp(std::string key) {
+    auto stream_iter = DnnOpFactory<T>::GetHashMap().find(key);
+    if (stream_iter == DnnOpFactory<T>::GetHashMap().end()) {
+      return nullptr;
+    } else {
+      return stream_iter->second;
+    }
+  }
+
+  void SetOp(std::string key, DnnOp* op) {
+    auto stream_iter = DnnOpFactory<T>::GetHashMap().find(key);
+
+    CHECK(stream_iter == DnnOpFactory<T>::GetHashMap().end());
+
+    DnnOpFactory<T>::GetHashMap()[key] = op;
+  }
+
+ private:
+  static inline std::unordered_map<std::string, DnnOp*> &GetHashMap() {
+    static thread_local std::unordered_map<std::string, DnnOp*> map_;
+    return map_;
+  }
+};
+
+// utility functions which convert int, double, bool or dims to string
```

Comment:
Please use StrCat() instead of rolling your own. Same everywhere.

Reply:
Fixed by using string "append" method

Reply:
Agree, some of functions are reserved for new future PR and should be removed.
Actually after creating this PR two weeks, we found an more efficient way to 
create "key". Let me replace ALL these method with the "efficient" way (from performance perspective).  Thanks!

Reply:
I have checked in a faster way in creating keys for DNN primitive reuse. Please help to review the change.  All "<Type>ToString() methods have been removed. Thanks!